### PR TITLE
ralberd/path dependent adjoint

### PR DESCRIPTION
**Pull request by ralberd**
Initial attempts at implementing the path-dependent adjoint sentitivity for the full FEA using J2 plasticity. Creating a draft PR so that I can get feedback as I flesh this out.

* Unit tests checking that the jax-computed gradients of internal variable updates match analytic values for small strain linear hardening
* "Unit test" with implementation of full path-dependent adjoint solve - Right now this doesn't test anything. I need to figure out how to check the gradient (finite differences?)
* I plan on adding more granular unit tests checking the individual parts of the adjoint equations
* I plan on figuring out a more efficient way of computing derivatives of internal variables as they are only dependent on quantities in their own element / integration point (computing for every other element / integration point is wasteful)

### CODE REVIEW

**Comment by ralberd** [comment]
There appears to be a bug in my adjoint system implementation for J2 plasticity. Feel free to take a look and see if anything sticks out. I will keep working on it this week.

**Comment by ralberd** [comment]
This is ready for review now. Let me know what you think!

File: optimism/inverse/test/FiniteDifferenceFixture.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+from optimism.test.MeshFixture import *
```
Comment (btalamini) [inline optimism/inverse/test/FiniteDifferenceFixture.py:1]: May I ask that you import the module directly and avoid `import *` in general? This pulls everythin into the top-level namespace makes it harder to see where things come from and to avoid name clashes. I realize that this idiom is used throughout optimism, but I've been trying to remove them bit by bit.

File: optimism/inverse/MechanicsInverse.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+from collections import namedtuple
+
+from optimism.JaxConfig import *
```
Comment (btalamini) [inline optimism/inverse/MechanicsInverse.py:3]: You shouldn't need this import. It's a legacy thing I'm tryin to remove. If it doesn't work without it, show me the error and I'll fix it.

File: optimism/inverse/MechanicsInverse.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                                       'ivs_update_jac_coords_vjp'])
+
+PathDependentResidualInverseFunctions = namedtuple('PathDependentResidualInverseFunctions',
+                                                  ['residual_jac_ivs_prev_vjp',
+                                                   'residual_jac_coords_vjp'])
+
+ResidualInverseFunctions = namedtuple('ResidualInverseFunctions',
+                                     ['residual_jac_coords_vjp'])
+
+def _compute_quadrature_point_field_gradient(u, shapeGrad):
```
Comment (btalamini) [inline optimism/inverse/MechanicsInverse.py:20]: This function appears to be copied and pasted from `FunctionSpace` - why not use that directly?

File: optimism/inverse/MechanicsInverse.py (hunk 2 above)
Comment (tupek2) [reply optimism/inverse/MechanicsInverse.py:3]: I happen to like import * from JaxConfig.  For me, its intent was to avoid having to constantly duplicate lots of lines like: from jax import grad, import blah, ... which I found to be annoying in every file.  It's an example where the DRY (dont repeat yourself) principal is in conflict with other principals.  For me, I like DRY here, but I know there is a trade off.  I understand that it pull in extra things, but that is essentially the point!  Without it, the tops of files are more verbose and it takes longer to get to the actual work of the file without such things.  If JaxConfig pull in things that may be duplicated or we don't want, we should delete them from JaxConfig.

Comment (btalamini) [reply optimism/inverse/MechanicsInverse.py:3]: I understand what you mean, but there are better ways of doing that. The wildcard import is discouraged pretty much universally because it makes it so hard to understand where things are coming from, and names can collide and unintentionally overwrite the one you want. It's essentially like a GOTO - you'd have to follow all the `import *` instances to see which `grad` you're ultimately getting.  Even the Python project itself [doesn't allow it in their style](https://peps.python.org/pep-0008/#imports) any more. If you want to use the concretization/substitution idea, there are ways to do it by importing things in the `__init__.py` file for the project, which is similar, but forces you to be explicit (like your  MatrixClass = Armadillo::Matrix example).

Comment (tupek2) [reply optimism/inverse/MechanicsInverse.py:3]: OK, can we basically put things like grad = jax.grad, etc. in __init__ then?  Basically things which are essentially always used and in the way we generally want to use them, so we can establish a standard way of doing it.  That was the intent of the .Config file.  In C++, the equivalent is basically #include "types.hpp", where types are put into the same namespace as the project on purpose.  That is essentially the same thing I want to do in python.  The issues is that putting it in __init__ puts in in EVERY file, where as "from optimism.types import *" is more selective.  So I don't really appreciate what the issue with that is.  We are not doing "from jax import *", we are including a types header.

Comment (btalamini) [reply optimism/inverse/MechanicsInverse.py:3]: I think we should discuss this offline. Ryan, you can leave things as they are until we come to an agreement.

File: optimism/inverse/MechanicsInverse.py (hunk 3 above)
Comment (ralberd) [reply optimism/inverse/MechanicsInverse.py]: Good catch, fixed it.

File: optimism/inverse/test/FiniteDifferenceFixture.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+from optimism.test.MeshFixture import MeshFixture
+from collections import namedtuple
+import numpy as np
```
Comment (btalamini) [inline optimism/inverse/test/FiniteDifferenceFixture.py:3]: For consistency, can you import numpy as `onp` here? We try to be consistent about calling the jax numpy `np` and the original numpy as `onp`. It makes it easier to understand a file at a glance.

File: optimism/Interpolants.py [hunk 5]
```diff
@@ -69,7 +69,7 @@ def get_lobatto_nodes_1d(degree):
     p = onp.polynomial.Legendre.basis(degree, domain=[0.0, 1.0])
     dp = p.deriv()
     xInterior = dp.roots()
-    xn = np.hstack(([0.0], xInterior, [1.0]))
+    xn = np.hstack((np.array([0.0]), xInterior, np.array([1.0])))
```
Comment (btalamini) [inline optimism/Interpolants.py:72]: Thanks, those warnings were annoying.

File: optimism/inverse/MechanicsInverse.py (hunk 2 above)
Comment (btalamini) [reply optimism/inverse/MechanicsInverse.py:3]: Ok, go ahead and leave the wildcard import. We'll make a sweeping change in the future to get rid of them. Try to leave out the JaxConfig import in the future.

**Review by btalamini — APPROVED** [review]
